src/cargo/core/package.rs

```diff
@@ -499,17 +499,18 @@ impl<'gctx> PackageSet<'gctx> {
         force_all_targets: ForceAllTargets,
     ) -> CargoResult<()> {
         fn collect_used_deps(
-            used: &mut BTreeSet<PackageId>,
+            used: &mut BTreeSet<(PackageId, CompileKind)>,
             resolve: &Resolve,
             pkg_id: PackageId,
             has_dev_units: HasDevUnits,
-            requested_kinds: &[CompileKind],
+            requested_kind: CompileKind,
```

Comment:
This seems mostly correct to me. Cargo allows only one target platform per artifact deps. Thanks folks. 👍🏾


With this RFC the approach here might still have some issues but it is not even landed yet so no need to worry
https://rust-lang.github.io/rfcs/3176-cargo-multi-dep-artifacts.html

Reply:
What we would do in that situation is just call `collect_used_deps` for each `target`.  This errs on the side of walking the tree multiple times, once for each target, rather than trying to keep state so that we only walk once.